lib/enigma_encoder.rb

```diff
@@ -1,3 +1,39 @@
 class EnigmaEncoder
 
+  def initialize
+    @whitelist = ('a'..'z').to_a << ' '
+  end
+
+  def keys(key)
+    key.split('').each_cons(2).map {|a,b| a + b}
+  end
+
+  def offset(date)
+    (date.to_i**2).to_s[-4..-1].split('')
+  end
+
+  def shifts(keys, offset)
+    keys.map.with_index do |key, index|
+      key.to_i + offset[index].to_i
+    end
+  end
+
+  def caesar_shift(character, shift)
+    index = @whitelist.index(character.downcase)
+    index ? @whitelist.rotate(shift)[index] : character
+  end
+
+  def encryption_info(encryption, key, date)
+    {encryption: encryption, key: key, date: date}
+  end
+
+  #possible to-do: make a "run_message_through_caesar_shifts" super method
+  def encrypt(message, key = nil, date = nil)
+    enigma_shifts = shifts(keys(key), offset(date)).rotate(-1)
+    encryption = message.split('').map do |character|
```

Comment:
Maps each character to its respective shifted character. The amount shifted comes from `shifts`, which cycles through its values for each character. Since the mapped value is the return value of the block, the rotate has to come before the `caesar_shift`. Alternatively, the character from `caesar_shift` could be set to a variable and returned after the rotate.